v4 to v5 Port

aeoranday opened the pull request:
This simply ports this repository from v4 to v5.

Most of the dependency for this port was data format related, which have already been ported.

Testing was done by viewing `t*_dump.py` for various files from both v4 and v5 and confirming the results are sensible.

This also features a slight change to docstrings and a missed enhancement in v4.

The usage of `process_tpstream` is consistent and was previously addressed when the naming change happened in `triggeralgs` for v5.

alessandrothea committed:
Adding tc making and rearranging configuration a bit

aeoranday committed:
Initial fragment removal.

aeoranday committed:
Bug fix.

aeoranday committed:
Prefer single row over single column.

aeoranday committed:
Removed fragment separation.

aeoranday committed:
Updated to make use of new TPData.tp_data definition.

aeoranday committed:
Updated to new format and added tick/second conversoin.

aeoranday committed:
Bug fix and conversion correction.

aeoranday committed:
Small style improvements.

aeoranday committed:
Fetches run number and file index.

aeoranday committed:
Shifted to double hist on PdfPages.

aeoranday committed:
Shifted to double hist on PdfPages.

aeoranday committed:
Detailed analysis on ADC Integral vs ADC Peak.

aeoranday committed:
Stops importing an empty binding.

aeoranday committed:
Merge pull request #11 from DUNE-DAQ/aeoranday/fragment-removal

Removed Fragment Separation From TAData & TPData Classes

aeoranday committed:
Merge branch 'production/v4' into aeoranday/removing-pybindsrc-ref.

aeoranday committed:
Merge pull request #15 from DUNE-DAQ/aeoranday/removing-pybindsrc-ref

Remove C Python Binding Import

aeoranday committed:
Flipped the linear/log ordering.

aeoranday committed:
Added seconds option, variable bins, and xlabel.

aeoranday committed:
Merge branch 'production/v4' into aeoranday/hdf5-file-attributes.

aeoranday committed:
Create dir on run id and file index and dump.

aeoranday committed:
Added an overwrite option and better word choice.

aeoranday committed:
--overwrite bug fix on mkdir logic.

aeoranday committed:
Abstract HDF5 reader class.

aeoranday committed:
Working TC data loading from raw HDF5.

aeoranday committed:
TC data member plotting.

aeoranday committed:
Unique save names and stats writing bug fix.

aeoranday committed:
Added time spans and ADC integral plots.

aeoranday committed:
Changed double scaling display order and improved xticks.

aeoranday committed:
Changed double scaling display order.

aeoranday committed:
Merge pull request #17 from DUNE-DAQ/aeoranday/hdf5-file-attributes

Using HDF5 File Attributes To Separate Outputs

aeoranday committed:
Expanding paths and style changes.

jcfreeman2 committed:
Syncing .github/workflows/dunedaq-v4-cpp-ci.yml

aeoranday committed:
Changed import style and comments.

aeoranday committed:
Corrected seconds/ticks unit change.

aeoranday committed:
Updated docs for new TC plots.

aeoranday committed:
Updated to use relative links.

aeoranday committed:
Merge branch 'production/v4' into aeoranday/tc-data-and-plots.

aeoranday committed:
Updated to better debug phrasing.

aeoranday committed:
Merge branch 'production/v4' into aeoranday/double-histogram.

aeoranday committed:
Cleaning some redundant info.

aeoranday committed:
Typo bug fix.

aeoranday committed:
Produces TCs. Creates at least 1 bad TC.

aeoranday committed:
Merge branch 'production/v4' into thea/tc_processing.

aeoranday committed:
Merge pull request #19 from DUNE-DAQ/aeoranday/tc-data-and-plots

TC Data Class & Plots

aeoranday committed:
Merge branch 'production/v4' into thea/tc_processing.

aeoranday committed:
Added a skip to saving empty frags.

aeoranday committed:
Merge branch 'production/v4' into aeoranday/double-histogram.

aeoranday committed:
Moved to script dir and made executable.

aeoranday committed:
Move histograms to 1 PDF and remove old plots.

aeoranday committed:
Added TA versions.

aeoranday committed:
Updated documentation.

aeoranday committed:
Updated help to include TCs.

aeoranday committed:
Merge pull request #21 from DUNE-DAQ/thea/tc_processing

Added TC Processing

aeoranday committed:
Merge pull request #22 from DUNE-DAQ/aeoranday/double-histogram

Updated TA & TP Plotting Style

aeoranday committed:
Created templated emulation unit class.

aeoranday committed:
Merge branch 'aeoranday/python-reader-refactoring' into aeoranday/python-styles.

aeoranday committed:
Added flake8 style configuration.

aeoranday committed:
Made concrete some common methods.

aeoranday committed:
Made fragment path filtering automatic and required.

aeoranday committed:
Changed naming and code structure to improve readability.

aeoranday committed:
Changed from --quiet to -v with levels.

aeoranday committed:
Changed coding style and plotting to single PDFs.

aeoranday committed:
Added a general PDF plotting class.

aeoranday committed:
Making use of PDFPlotter in dump scripts.

aeoranday committed:
Added common useful function methods.

aeoranday committed:
Added a pip requirements.txt.

aeoranday committed:
Added documentation on the Python trgtools module.

aeoranday committed:
Added more details on Python trgtools module.

aeoranday committed:
Removed a debug comment.

aeoranday committed:
Updated for the new save format.

aeoranday committed:
Merge pull request #24 from DUNE-DAQ/prep-release/dunedaq-v4.3.0

Prep release/dunedaq v4.3.0

aeoranday committed:
Merge pull request #23 from DUNE-DAQ/aeoranday/python-styles

Python Style Update

aeoranday committed:
Making use of the emulation units.

aeoranday committed:
Added a naive timing for TA and TC processes.

aeoranday committed:
Changed to relative start times.

aeoranday committed:
Merge pull request #27 from DUNE-DAQ/aeoranday/event-display-relative-time

Changed Event Displays To Relative Start Times

aeoranday committed:
Adding a check for 0 TA/TC creation.

aeoranday committed:
Fixed shared Python data members.

aeoranday committed:
Merge pull request #28 from DUNE-DAQ/aeoranday/python-data-sharing-fix

Fixed Shared TXReader Data Members

aeoranday committed:
Updated documentation.

aeoranday committed:
Merge pull request #29 from DUNE-DAQ/aeo/update-process-tpstream-md

Updated Emulator Documentation

aeoranday committed:
Added timing measures for TX into TY.

aeoranday committed:
Added find_dependency.

aeoranday committed:
Merge pull request #31 from DUNE-DAQ/aeo/find-dependency

Added find_dependency

aeoranday committed:
Merge branch 'production/v4' into aeoranday/emulation-units.

aeoranday committed:
Merge branch 'production/v4' into aeoranday/emulation-units.

aeoranday committed:
Refined timing measure to creator TP/TA.

aeoranday committed:
Merge pull request #26 from DUNE-DAQ/aeoranday/emulation-units

Process Emulation Units & TA/TC Process Timing

aeoranday committed:
Updating to v1.1.0.

aeoranday committed:
Merge pull request #32 from DUNE-DAQ/aeo/v1.1.0

Updating to v1.1.0.

aeoranday committed:
Fixed multi-fragment TimeSlice processing.

aeoranday committed:
Merge pull request #33 from DUNE-DAQ/aeo/process-tpstream-multi-frag-timeslice

Fixed multi-fragment TimeSlice processing.

MRiganSUSX committed:
Updated version to v1.1.1

MRiganSUSX committed:
Merge pull request #34 from DUNE-DAQ/mrigan/version_update

Updated version to v1.1.1

Commit:
Syncing .github/workflows/dunedaq-v4-cpp-ci.yml

jcfreeman2 committed:
Merge pull request #35 from DUNE-DAQ/prep-release/fddaq-v4.4.0

Prep release/fddaq v4.4.0

andrewmogan committed:
Syncing .github/workflows/dunedaq-v4-cpp-ci.yml

ArturSztuc committed:
Adding latency option to the emulation app, + more quiet

ArturSztuc committed:
More robust csv latencies filename/path for saving

ArturSztuc committed:
Adding full latency output per-tp in emulation

ArturSztuc committed:
Script that converts large tp latency output to smaller TA latencies

Also added new trgtools dependency on pandas

ArturSztuc committed:
Updating and adding the READMEs for latency measurements

ArturSztuc committed:
Making latencies-gen python script executable

ArturSztuc committed:
Removing unused variables from tplatencies python script

ArturSztuc committed:
Merge pull request #36 from DUNE-DAQ/feature/ta_latencies

TA emulation latencies

ArturSztuc committed:
Updating version to 1.1.2

ArturSztuc committed:
Updating version to 1.1.2

ArturSztuc committed:
Merge pull request #37 from DUNE-DAQ/asztuc/version_update

Version update to v1.1.2

jcfreeman2 committed:
Merge pull request #38 from DUNE-DAQ/patch/fddaq-v4.4.x

Patch/fddaq v4.4.x

andrewmogan committed:
Syncing .github/workflows/dunedaq-v4-cpp-ci.yml

aeoranday committed:
Use trgdataformats Algorithms and Types.

aeoranday committed:
Use trgdataformats TC Py-capsule.

aeoranday committed:
Add Dynamic Algorithm and Type Plotting.

aeoranday committed:
Merge pull request #40 from DUNE-DAQ/aeo/tc-data-update

Use trgdataformats TC Py-capsule.

aeoranday committed:
Fix typo.

aeoranday committed:
Add No TX Fragment Guards.

Added a check that a file contains TX fragments before processing.

ArturSztuc committed:
Merge pull request #41 from DUNE-DAQ/aeo/enum-update

trgdataformats Enum Update Propagation

aeoranday committed:
Add Alg/Type Histogram Fontsize.

Added a fontsize for the awkward, long names in the alg/type histograms.

aeoranday committed:
Squash Alg/Type Histogram Bins.

Made alg/type histogram bins slightly smaller for there to be a gap.

aeoranday committed:
Update To v1.1.3.

aeoranday committed:
Merge pull request #42 from DUNE-DAQ/aeo/v1.1.3-bump

Version Bump To v1.1.3

jcfreeman2 committed:
Merge pull request #43 from DUNE-DAQ/patch/fddaq-v4.4.x

Patch/fddaq v4.4.x

aeoranday committed:
Add Color To Hist Axes.

Added linear and log colors to the corresponding axes.

aeoranday committed:
Add Unity And Transparency.

Added the unity line and made markers transparent.

aeoranday committed:
Correct Help Descriptions

aeoranday committed:
Merge pull request #44 from DUNE-DAQ/aeo/style-update

Plotting Style Update

ArturSztuc committed:
Bumping version to v1.1.4

ArturSztuc committed:
Merge pull request #45 from DUNE-DAQ/asztuc/version_bump_v1.1.4

Bumping version to v1.1.4

jcfreeman2 committed:
Merge pull request #47 from DUNE-DAQ/patch/fddaq-v4.4.x

Patch/fddaq v4.4.x

aeoranday committed:
Merge branch 'production/v4' into aeo/v4_to_v5.

aeoranday committed:
Update TA Types/Algs and Correct Docstrings.

MRiganSUSX commented:
Hi Alex,

Thanks for the work on this. Here is my first round of review: 
```
new tag for v5
fix build warnings
link to process_tpstream documentation (same as for the others)
note on what channel map is for PD2HD
the default config for process_tpstream has invalid names for the algos...

process_tpstream:
"
 TS gap: 62499456 999.991296 ms
		Average TA Time Process (0 TAs): 0 ns.
		Average TC Time Process (0 TCs): 0 ns.
"

for process_tpstream --latencies please add a header for columns

for ta_dump, could the list of commands at the bottom be either ordered alphabetically or in the order they were introduced in the text above

seen an error with ta_dump in v4:
"
(dbt) [mrigan@np04-srv-016 trgtools_old]$ ta_dump.py /nfs/rscratch/mrigan/np04hd_raw_run029349_0000_dataflow0_datawriter_0_20240926T114848.hdf5 
Saving outputs to ./ta_29349-0000_figures_0000.*
Number of TAs: 3
Writing descriptive statistics to ta_29349-0000_figures_0000.txt.
Traceback (most recent call last):
  File "/nfs/home/mrigan/dune-daq/fddaq-v4.4.8-a9/install/trgtools/bin/ta_dump.py", line 551, in <module>
    main()
  File "/nfs/home/mrigan/dune-daq/fddaq-v4.4.8-a9/install/trgtools/bin/ta_dump.py", line 488, in main
    plot_data = np.array([datum.value for datum in data.ta_data[ta_key]], dtype=int)
  File "/nfs/home/mrigan/dune-daq/fddaq-v4.4.8-a9/install/trgtools/bin/ta_dump.py", line 488, in <listcomp>
    plot_data = np.array([datum.value for datum in data.ta_data[ta_key]], dtype=int)
AttributeError: 'numpy.uint8' object has no attribute 'value'
"
this was not the case for the same file for v5 (for same file). it makes it impossible to compare 'apples to apples'

(tx_dump) nit-picking, but if you have time:
  -- detector IDs are integers (plot)
  -- number of TAs per TC does not need to go below 0 (same for other counting plots)
  -- 'TC Relative Time Spans' plot: red bars need to be on the top of the black line + screenshot
  -- (tp_dump)'TP Time Over Threshold vs Channel': little higher transparency on the dots

TP's ADC peak > ADC integral: np04hd_raw_run029194_0002_dataflow2_datawriter_0_20240916T072613.hdf5 + screenshot

convert_latencies:
"
(dbt) [mrigan@np04-srv-016 trgtools_old]$ convert_tplatencies.py -r /nfs/rscratch/mrigan/np04hd_raw_run029349_0000_dataflow0_datawriter_0_20240926T114848.hdf5 -l ta_timings_processed.csv 
Loading TAs from hdf5
Loading TPs from csv
Traceback (most recent call last):
  File "/nfs/home/mrigan/dune-daq/fddaq-v4.4.8-a9/install/trgtools/bin/convert_tplatencies.py", line 109, in <module>
    main()
  File "/nfs/home/mrigan/dune-daq/fddaq-v4.4.8-a9/install/trgtools/bin/convert_tplatencies.py", line 92, in main
    latencies_whole = tp_data_latencies.iloc[first_tp_idx + 1:tp_making_ta_idx + 1,2].sum()
TypeError: unsupported operand type(s) for +: 'NoneType' and 'int'
"
both v4 and v5...

in py-trgtools.md:
  -- Plotting -> Example: the example script is broken (doesn't load tp_data, when plotting it's not accessing the data correctly)
```

aeoranday committed:
Clean Compiler Warnings.

aeoranday committed:
Update Docs.

Also included a link to the channel maps table.

aeoranday committed:
Add Latency CSV Headers.

aeoranday committed:
Change Flag Order.

aeoranday committed:
Use Integer X-Ticks.

Added and using an option to use integer x-ticks in PDFPlotter histograms.

aeoranday committed:
Increase Errorbar Marker Size.

Previously invisible.

aeoranday committed:
Fix Minor Tick Coloring.

Log minor ticks were not being colored correctly.

aeoranday committed:
Add TP Transparency.

aeoranday committed:
Remove Whitespace.

aeoranday committed:
Remove Unused Option.

Users should define which maker/plugin in a config JSON.

aeoranday committed:
Link Libraries.

._.

aeoranday committed:
Add Missing Edge Case & Remove Warning.

The last TP is sometimes missing, so skip it. Removed a pandas deprecation warning.

aeoranday committed:
Replace 'np.ndarray' With NDArray.

This is more correct to PEP 484.

aeoranday committed:
Add Primary Data Indexing.

Indexing the reader class now returns the indexed content of the primary data.

aeoranday committed:
Update TX Extraction.

Follows the flow change that happened in `trigger`.

aeoranday committed:
Upgrade to v2.0.0.

ArturSztuc commented on apps/tapipe.cxx:
```diff
@@ -245,18 +245,18 @@ main(int argc, char* argv[])
+  char* payload = static_cast<char*>(malloc(payload_size));
 
 
   size_t offset(0);
   for ( const auto& ta : ta_buffer ) {
-    triggeralgs::write_overlay(ta, payload+offset);
+    triggeralgs::write_overlay(ta, static_cast<void*>(payload+offset));
     offset += triggeralgs::get_overlay_nbytes(ta);
   }
 
-  daqdataformats::Fragment ta_frag(payload, payload_size);
+  daqdataformats::Fragment ta_frag(static_cast<void*>(payload), payload_size);
 
-  free(payload);
+  free(static_cast<void*>(payload));
```
Both here and in `EmulationUnit`, I would totally avoid using `malloc` and `free`. This is low-level C that should only be used if really cannot be avoided. In this case we could simply do:

```
  std::vector<char> payload(payload_size);

  size_t offset(0);
  for ( const auto& ta : ta_buffer ) {
    triggeralgs::write_overlay(ta, static_cast<void*>(payload.data()+offset));
    offset += triggeralgs::get_overlay_nbytes(ta);
  }
  daqdataformats::Fragment ta_frag(static_cast<void*>(payload.data()), payload_size);
```

And no `free` would be needed, since vector would go out of scope anyway (but could clear it if you must).


BUT this is not related to v4-to-v5 movement, so I'm fine if that's done at later date with a separate PR

ArturSztuc commented on include/trgtools/detail/EmulationUnit.hxx:
```diff
@@ -80,18 +80,19 @@ EmulationUnit<T, U, V>::emulate_vector(const std::vector<T>& inputs) {
 
-  void* payload = malloc(payload_size);
+  // Awkward type conversion to avoid compiler complaints on void* arithmetic.
+  char* payload = static_cast<char*>(malloc(payload_size));
```
ditto re. malloc & free, could use vector here instead.

ArturSztuc commented on apps/process_tpstream.cxx:
```diff
@@ -212,6 +212,7 @@ int main(int argc, char const *argv[])
   if (latencies) {
     std::filesystem::path output_path(output_file_path);
     ta_emulator->set_timing_file((output_path.parent_path() / ("ta_timings_" + output_path.stem().string() + ".csv")).string());
+    ta_emulator->write_csv_header("TP Time Start,TP ADC Integral,Time Diffs,Is Last TP In TA");
```
I like this, although I wish it was done in a separate PR, not in a simple "v4 to v5". I'm now having to do PR review commit-by-commit :(

aeoranday replied:
I understand the out-of-scopeness, but it was part of the requested changes. I tried to keep the commits well separated for what they changed.

ArturSztuc commented on python/trgtools/TAReader.py:
```diff
@@ -7,6 +7,7 @@
 import trgdataformats
 
 import numpy as np
+from numpy.typing import NDArray
```
Had no idea numpy has its own typing module. Nice.

ArturSztuc reviewed (approved):
LGTM. 
One suggestion made, but it's not related to v4->v5 movement so fine to ignore.

aeoranday committed:
Remove Placeholder Unit Test.

Only removes from building. The cxx is kept for reference.

MRiganSUSX reviewed (commented):
Thanks @aeoranday for improvements, it's already better.

Here are more comments after another round of review:

1. running `process_tpstream` over the same tpstream hd5 file, same configuration (barring change in algo names, but both prescale):
V4:
```
		Average TA Time Process (19702 TAs): 7410 ns.
		Average TC Time Process (197 TCs): 7758 ns.
-- Finished TSL 118490:0
```
V5:
```
		Average TA Time Process (19695 TAs): 18111 ns.
		Average TC Time Process (197 TCs): 21811 ns.
-- Finished TSL 118490:0
```
=> different number of TAsm and much much slower?
(output file also very slightly smaller - guessing that's the 'missing' TAs ?)

2. running `process_tpstream` with `--latencies`:
-- (same issue as above with missing TAs)
-- the 'Time Diffs' parameters differ for the same TA(s):
V4:
```
108017674586074187,16506,7949,1   <- ??
108017674565021675,26052,147,0
```
V5:
```
108017674586074187,16506,8115,0
108017674565021675,26052,953,0
```

3. `convert_tplatencies`:
what exactly is the point of convert_tplatencies? It does not seem that there any scripts available that would use the output of this. And, if what it produces is the desired result, why don't we get it straight from tp_process with --latencies? is that inbetween output used for anything else other than the input to this converter ? 

4. `tp_dump`:
still get weird TPs (attach ss)
![Screenshot from 2024-10-14 13-26-26](https://github.com/user-attachments/assets/155c5249-fcfa-4ee5-a26e-ff86b60854e5)

5. `tc_dump`:
- when using on the output file made with `process_tpstream` (the input to that process step is the same in both versions):
even though the # of TCs made IS the same, the TCs are different, as the TC stats are different:
V4:
```
time_candidate
Reference Statistics:
        Total # TCs = 10,
        Mean = 2.13,
        Std = 1.37,
        Min = 0.0,
        Max = 4.220520496368408.
Anomalies:
        # of >3 Sigma TCs = 0,
        # of >2 Sigma TCs = 0.


time_end
Reference Statistics:
        Total # TCs = 10,
        Mean = 2.13,
        Std = 1.37,
        Min = 0.0,
        Max = 4.220515489578247.
Anomalies:
        # of >3 Sigma TCs = 0,
        # of >2 Sigma TCs = 0.


time_start
Reference Statistics:
        Total # TCs = 10,
        Mean = 2.13,
        Std = 1.37,
        Min = 0.0,
        Max = 4.220520496368408.
Anomalies:
        # of >3 Sigma TCs = 0,
        # of >2 Sigma TCs = 0.
```
V5:
```
time_candidate
Reference Statistics:
        Total # TCs = 10,
        Mean = 2.02,
        Std = 1.46,
        Min = 0.0,
        Max = 4.214995384216309.
Anomalies:
        # of >3 Sigma TCs = 0,
        # of >2 Sigma TCs = 0.


time_end
Reference Statistics:
        Total # TCs = 10,
        Mean = 2.02,
        Std = 1.46,
        Min = 0.0,
        Max = 4.214994430541992.
Anomalies:
        # of >3 Sigma TCs = 0,
        # of >2 Sigma TCs = 0.


time_start
Reference Statistics:
        Total # TCs = 10,
        Mean = 2.02,
        Std = 1.46,
        Min = 0.0,
        Max = 4.214995384216309.
Anomalies:
        # of >3 Sigma TCs = 0,
        # of >2 Sigma TCs = 0.
```
(using this dump on the same raw hdf5 (without the process_ step), the output is identical).

6. `py-trgtools`:
-- very last command block (Plotting->Example) is missing command to load data: 
tpr.read_all_fragments()            (or something similar that takes subset)
without that:
```
>>> print(tpr['adc_peak'])
[]
```
-- the output example.pdf seems to be broken / corrupted (empty) file (tried 3 different inputs)
-- there is an error message when exiting the interactive python session:
```
Exception ignored in: <function PDFPlotter.__del__ at 0x7fb183882680>
Traceback (most recent call last):
  File "/nfs/home/mrigan/dune-daq/v5.2-241010/install/trgtools/lib64/python/trgtools/plot/PDFPlotter.py", line 91, in __del__
  File "/nfs/home/mrigan/dune-daq/v5.2-241010/.venv/lib/python3.10/site-packages/matplotlib/backends/backend_pdf.py", line 2721, in close
  File "/nfs/home/mrigan/dune-daq/v5.2-241010/.venv/lib/python3.10/site-packages/matplotlib/backends/backend_pdf.py", line 828, in finalize
  File "/nfs/home/mrigan/dune-daq/v5.2-241010/.venv/lib/python3.10/site-packages/matplotlib/backends/backend_pdf.py", line 974, in writeFonts
  File "/nfs/home/mrigan/dune-daq/v5.2-241010/.venv/lib/python3.10/site-packages/matplotlib/backends/backend_pdf.py", line 1451, in embedTTF
  File "/nfs/home/mrigan/dune-daq/v5.2-241010/.venv/lib/python3.10/site-packages/matplotlib/backends/backend_pdf.py", line 1182, in embedTTFType3
ImportError: sys.meta_path is None, Python is likely shutting down
```

I understand that lot (if not all) of these are not caused by your port. I'm just listing my findings here as I was going through testing. 
Please, for those points that you are not fixing for this PR, make issues.
